code_generation/producer.py

```diff
@@ -1,16 +1,28 @@
-import code_generation.quantity as q
+from __future__ import annotations  # needed for type annotations in > python 3.7
+
 import logging
+from typing import Any, Dict, List, Set, Union
+from code_generation.exceptions import InvalidProducerConfigurationError
+
+import code_generation.quantity as q
 
 log = logging.getLogger(__name__)
 
 
-class SafeDict(dict):
-    def __missing__(self, key):
+class SafeDict(Dict[Any, Any]):
+    def __missing__(self, key: Any) -> Any:
         return "{" + key + "}"
 
 
 class Producer:
-    def __init__(self, name, call, input, output, scopes):
+    def __init__(
+        self,
+        name: str,
+        call: str,
+        input: Union[List[q.Quantity], Dict[str, List[q.Quantity]]],
```

Comment:
The input to a Producer can also be from nanoAOD, so can `q.Quantity` and `q.NanoAODQuantity` be seen as the same? Or is this even relevant here?

Reply:
A `NanoAODQuantity` is a `Quantity` since  the `NanoAODQuantity`class inherits from `Quantity`, so `q.Quantity` covers both cases